### ISSUE TITLE
Make the maximum number of listeners in Gateway CRD configurable

### ISSUE THREAD

**Issue by bravenut** [kubernetes-sigs/gateway-api#2869]
**What would you like to be added**:
I would like the gateway CRD template to be improved by making the `maxItems` on the gateway listeners to be configurable instead of it being hard coded to be 64 items (see [experimental gateway CRD](https://github.com/kubernetes-sigs/gateway-api/blob/main/config/crd/experimental/gateway.networking.k8s.io_gateways.yaml#L1099C17-L1099C29)).
 
**Why this is needed**:
Our current setup uses an "edge" gateway (Kubernetes Gateway API gateway) routing requests to other "inner" gateways based on the path prefix. We currently have around 30 inner gateways which the edge gateway routes requests to, but we will have more in the future, so we will hit the hard coded limit of 64 max really soon.

**Comment by howardjohn** [comment]
There is a lot of discussion around similar topics in https://github.com/kubernetes-sigs/gateway-api/pull/1863.

IMO, I strongly agree with the need to have >64. But a few notes

* Just raising the limit is probably not sufficient, as realistically its not going to be so on the order of the largest users requirements (lets say 10k listeners)
* A more scalable approach is to split out the listeners, as the merging GEP above does
* Having different limits in different CRDs is probably a bad precedent to set; there should be 1 single API definition across users/clusters/implementations
 
cc @robscott

**Comment by youngnick** [comment]
Hi @bravenut, I don't really understand how you're doing this, since a Gateway is not a valid `targetRef` in any implementation I know about.

The intended design here is for the Gateway to hold _summarized_ Listeners (most likely bounded by a wildcard certificate or something, so you could have two listeners, one for `*.example.com` and one for `*.example.net`, and then have HTTPRoutes claim more specific domains beneath those listeners, using `morespecific.example.com` to attach to only the `*.example.com` listener, for example).

We'd need some example config of what you want to do here to understand your request better. Remember that Listeners are only required to be distinct on the following fields:
- port
- protocol
- TLS details
- optionally Hostname

**Comment by kghost** [comment]
> The intended design here is for the Gateway to hold _summarized_ Listeners (most likely bounded by a wildcard certificate or something, so you could have two listeners, one for `*.example.com` and one for `*.example.net`, and then have HTTPRoutes claim more specific domains beneath those listeners, using `morespecific.example.com` to attach to only the `*.example.com` listener, for example).

Sometimes, they can not be summarized, for example, to build a CDN, we probably need hundreds of thousands listeners, with different TLDs.

**Comment by youngnick** [comment]
I think that this use case is probably better addressed with something like #3213 - although I still think that having hundreds of thousands of listeners on a single Gateway is way outside what our current scaling design takes into account, so it's likely that there will be scale problems we haven't considered.

Also, if you're terminating TLS or otherwise wanting to do inspection on the HTTP metadata at the Gateway implementation point, then using `hostname` in HTTPRoute and attaching lots of HTTPRoutes to a single Listener is probably a slightly better design.

More generally, it's generally going to be way more scalable to attach lots of HTTPRoutes to a Gateway than it is to have lots of Listeners (if you can easily summarize the listeners on the same port or something).

**Comment by woodgear** [comment]
notice that due the exist of x-kubernetes-validations, you could not just remove the limit of maxItems, or it will report error like 
```
The CustomResourceDefinition "gateways.gateway.networking.k8s.io" is invalid:
* spec.versions[0].schema.openAPIV3Schema.properties[spec].properties[listeners].x-kubernetes-validations[3].rule: Forbidden: estimated rule cost exceeds budget by factor of 4.6x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
* spec.versions[0].schema.openAPIV3Schema.properties[spec].properties[listeners].x-kubernetes-validations[4].rule: Forbidden: estimated rule cost exceeds budget by factor of 9.3x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
* spec.versions[0].schema.openAPIV3Schema.properties[spec].properties[listeners].x-kubernetes-validations[4].rule: Forbidden: contributed to estimated rule cost total exceeding cost limit for entire OpenAPIv3 schema
* spec.versions[0].schema.openAPIV3Schema.properties[spec].properties[listeners].x-kubernetes-validations[3].rule: Forbidden: contributed to estimated rule cost total exceeding cost limit for entire OpenAPIv3 schema
* spec.versions[0].schema.openAPIV3Schema: Forbidden: x-kubernetes-validations estimated rule cost total for entire OpenAPIv3 schema exceeds budget by factor of 1.388001x (try simplifying the rule, or adding maxItems, maxProperties, and maxLength where arrays, maps, and strings are declared)
```
